web/server/go.md

```diff
@@ -0,0 +1,498 @@
+is due for a refactor.
+
+Unit testing handlers often involves mocking
+[service layers](#services).
+From there,
+they will answer the question of:
+"given a set of service layer responses,
+what response does the handler return?"
+This verifies the response/request responsibilities of the handler,
+and can be used to test edge cases with little test performance impact.
+
+### Models
+
+Put simply, models describe data.
+They are agnostic to the data source
+(JSON over HTTP, in-memory data structures, DB tables, etc.),
+however in many cases the term will be used synonymously
+with data queried from a DB.
+
+The model code should provide the structure of the data
+along with methods dictating how outside sources interact with it.
+
+For example,
+here's a pretty simple example of a model struct for a dog.
+
+```go
+package models
+
+type Dog struct {
+    ID      uuid.UUID `json:"id" db:"id"`
+    Name    string    `json:"name" db:"name"`
+    Owner     Owner
+}
+```
+
+It has a a unique ID, name, and `Owner`.
+The latter being a `struct` that encompasses an association.
+
+The field tags follow the syntax patterns for the
+[json](https://golang.org/pkg/encoding/json/)
+and [sqlx](https://godoc.org/github.com/jmoiron/sqlx) packages.
+This is not to prescribe specific packages,
+but to provide a more decorative example that may be encountered in a Truss codebase.
+
+Interacting with this model is likely going to happen in two ways:
+
+1. marshalling from an outside data source
+2. modifying/viewing the data in Go
+   (ex. executing business logic).
+
+#### Marshalling Models
+
+In the most basic cases of accepting an HTTP request
+or writing/reading from a SQL DB,
+marshalling is unlikely to happen manually.
+Many web app frameworks or DB connectors/ORMs
+are going to handle that hidden by their public APIs
+and provide convenience query methods
+or, in the case of JSON HTTP requests,
+parameter wrappers for handlers.
+Within a handler or service,
+the model is likely to be accessed only in its Go struct form.
+If you're curious what that marshalling code might look like,
+you can look at this [json blog post](https://blog.golang.org/json-and-go)
+or take a look at some of the
+[sqlx column mapping code](https://github.com/jmoiron/sqlx/blob/master/sqlx.go#L23)
+that works in tandem with the sql standard library.
+We'll get more into the interaction between those packages
+and our application code in the next section about [services](#services).
+
+#### Working with Models in Go
+
+In the second case,
+we need to utilize our models to store results of our application logic.
+After all, model structs and databases don't deliver client value on their own.
+Let's take a look back at our dog model
+and perform some operations.
+
+First, let's try to add a dog.
+
+```go
+package models
+
+type Dog struct {
+    ID      uuid.UUID `json:"id" db:"id"`
+    Name    string    `json:"name" db:"name"`
+    Owner     Owner
```

Comment:
I'm open to being shown a better way, but in my experience, fully including another model object (`Owner`) in this model object (`Dog`) makes more advanced usages harder. I've learned to prefer if there were just a simple `OwnerID`... This makes it possible for the `Owner` to actually exist in a different data store, and for `Owner`s and `Dog`s to be cached independently.

Super open to push back on this... I could well be over-complicating and over-engineering and over-future-proofing.

Reply:
How would someone do a join query in this case? Would the relations live in different structs or arrays? How would it get translated to nested JSON?

Reply:
Not sure I'm following on the JOIN query bit... (Probably because I've been mostly in NoSQL for the last couple of years...) If you've got the `Dog.ID` and the `OwnerID`, you have everything you need, right?
As for "where" the relations live, it depends on the cardinality relations I believe. I was assuming one `Owner` has many `Dog`s, meaning the list of dogs for an owner is just an array of `ID`s... And a `Dog` has 0-1 `Owner`, which lives at `Dog.OwnerID`.
Translating to JSON would live in the "transport" layer... But I'd move away from needing to represent it that way... Opting more for a "each noun is individually retrievable (and cacheable), let the front-end and/or GraphQL sort 'em out" kind of approach.

This is probably too complex and confusing of a hand-waving thing I'm doing for this venue. Hrm... 🤔

Reply:
Maybe I should take the "if you can't explain it well enough, you don't really understand it yourself" counsel... 😀